### PR TITLE
DMatrix transition for NumPy\Pandas

### PR DESCRIPTION
**Pull request by etveritas**
DMatrix transition from python numpy or pandas.

### CODE REVIEW

**Comment by etveritas** [comment]
@aksnzhy Do you think this design is feasible?

**Comment by aksnzhy** [comment]
This PR looks nice. I will review this code as soon as possible.

**Comment by aksnzhy** [comment]
@etveritas Hi, could you please add some demo by using this API? For example, we can add each demo (`demo` path) a python example by using this interface. And also, we need to update the document `python_package.rst` to introduce this interface. Thanks for the effort!

**Comment by etveritas** [comment]
@aksnzhy No problem. Before it, I'll enchance this PR, like more expectional handles.

**Comment by etveritas** [comment]
@aksnzhy All finished for this PR.
My thought of this design is add two hyper parameters ``from_file``, ``res_out`` to indicate DMatrix trainsition and result for numpy respectively, and this two parameters is invisible for CLI and indirectly use for users who use forepart language. If users tranform numpy/pandas to DMatrix, we set DMatrix dependent instead ``hash value``(not use ). There are some features simple but not implement, just want to observe users demand, if need will add for other PR or version.

**Comment by aksnzhy** [comment]
@etveritas I'm sorry, actually I review this code 15 days ago and I wrote the comments. But I didn't click the submit bottom to submit it...

File: python-package/xlearn/data.py [hunk 1]
```diff
@@ -14,5 +14,103 @@
...
-    pass
\ No newline at end of file
+    def __init__(self, data, label=None, field_map=None):
+        """
+        Initial function.
+        Parameters:
+        data: NumPy 2D or pandas DataFrame of features data.
+        label: one-dimentional array, it presents samples label.
+        field_map: one-dimentional array, it presents the feautres'field repsepctively.
+        This field_map like, [1, 2, 1, 3] means, the first and third features belong to field one, and the second belongs to field two, and so on.
```
[the remark this answers is not in the document]
Comment (etveritas) [reply python-package/xlearn/data.py]: No, not include in this PR

File: doc/api_list.rst [hunk 2]
```diff
@@ -162,6 +178,7 @@ API List: ::
     model.disableLockFree()   # Disable lock-free training.
 
     model.disableEarlyStop()  # Disable early-stopping.
+    
 
 Parameter List: ::
```
[the remark this answers is not in the document]
Comment (etveritas) [reply doc/api_list.rst]: Already pushed.

File: doc/python_package.rst [hunk 3]
```diff
@@ -529,6 +529,66 @@ the training, and it just train the model quietly: ::
...
+    # that is, you can also pass xl.DMatrix for this API now
+    fm_model.setTest(xdm_test)  # Test data
+
+    # Start to predict
+    # The output result will be stored in output.txt
+    # if no result out path setted, we return res as numpy.ndarray
+    res = fm_model.predict("./model_dm.out")
+
+**Note:** Train from DMatrix is not support cross validation now, and we will add this feature soon later. 
+
```
[the remark this answers is not in the document]
Comment (etveritas) [reply doc/python_package.rst]: Already pushed.

**Comment by etveritas** [comment]
- [X] Not skip zero when read CSV.

- [X] Add DMatrix transition from numpy\pandas.

- [X] Add some demos for DMatrix transition(both regression and classification).

- [X] Correct the xLearn Error handle.

**Comment by aksnzhy** [comment]
> This is the task list about this PR
> 
> * [x]  Not skip zero when read CSV(#170).
> * [x]  Add DMatrix transition from numpy\pandas.
> * [x]  Add some demos for DMatrix transition(both regression and classification).
> * [x]  Correct the xLearn Error handle for `c_api`.
> 
> @aksnzhy And I have a question about this code in file [fm_score.cc](https://github.com/aksnzhy/xlearn/blob/master/src/score/fm_score.cc#L57)
> 
> ```c++
> std::vector<real_t> sv(aligned_k, 0);
> real_t* s = sv.data();
> ...
> __m128 XMMs = _mm_load_ps(s+d);
> ```
> 
> Is `s` returned from `std::vector::data` the aligned address automatic?

I don't think this memory is aligned.

**Comment by aksnzhy** [comment]
@etveritas Thanks for this effort! I will merge this PR. Also, could you please update the document in `xlearn_doc` and `xlearn_doc_cn`, and the website will automatically updated when this two Repo changed.

**Comment by aksnzhy** [comment]
I find the result of 'run_titanic_no_cv_pandas.py' and 'run_titanic_no_cv.py' is quite different. Could you please check this problem?

**Comment by etveritas** [comment]
@aksnzhy Yeah, I have already made PRs for this two repos.

**Comment by etveritas** [comment]
@aksnzhy Without lock-free?

**Comment by aksnzhy** [comment]
> @aksnzhy I test the correctness of DMatrix trainsition without lock-free.

Oh I see. You are right.

**Comment by etveritas** [comment]
> > @aksnzhy I test the correctness of DMatrix trainsition without lock-free.
> 
> Oh I see. You are right.

Haha, I added explaination for users in these files.

**Comment by etveritas** [comment]
> > This is the task list about this PR
> > 
> > * [x]   Not skip zero when read CSV(#170).
> > * [x]   Add DMatrix transition from numpy\pandas.
> > * [x]   Add some demos for DMatrix transition(both regression and classification).
> > * [x]   Correct the xLearn Error handle for `c_api`.
> > 
> > @aksnzhy And I have a question about this code in file [fm_score.cc](https://github.com/aksnzhy/xlearn/blob/master/src/score/fm_score.cc#L57)
> > ```c++
> > std::vector<real_t> sv(aligned_k, 0);
> > real_t* s = sv.data();
> > ...
> > __m128 XMMs = _mm_load_ps(s+d);
> > ```
> > 
> > 
> > Is `s` returned from `std::vector::data` the aligned address automatic?
> 
> I don't think this memory is aligned.

@aksnzhy That's my point, I find ``_mm_load_ps`` must receive aligned address in [Intel Intrinsics Guide](https://software.intel.com/sites/landingpage/IntrinsicsGuide/#text=_mm_load_ps&expand=3297), but it works when compile...

**Pull request merged**